test/unit/org/apache/cassandra/utils/Retry.java

```diff
@@ -0,0 +1,222 @@
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        SCHEDULED.schedule(() -> run0(executor, future, fn), nanos, TimeUnit.NANOSECONDS);
+        return future;
+    }
+
+    private static void run0(final Executor executor, final CompletableFuture<Void> future, final Runnable fn)
+    {
+        try
+        {
+            executor.execute(() -> {
+                try
+                {
+                    fn.run();
+                    future.complete(null);
+                }
+                catch (Exception e)
+                {
+                    future.completeExceptionally(e);
+                }
+            });
+        }
+        catch (Exception e)
+        {
+            future.completeExceptionally(e);
+        }
+    }
+
+    /**
+     * Continously attempting to call the provided future supplier until successful or until no longer able to retry.
+     *
+     * @param maxRetries to allow
+     * @param fn asyncronous operation to retry
+     * @param retryableException used to say if retry is allowed
+     * @return future representing the result.  If retries were not able to get a successful result, the exception is the last exception seen.
+     */
+    public static <A> CompletableFuture<A> retryWithBackoff(final int maxRetries,
+                                                            final Supplier<CompletableFuture<A>> fn,
+                                                            final Predicate<Throwable> retryableException)
+    {
+        CompletableFuture<A> future = new CompletableFuture<>();
+        retryWithBackoff0(future, 0, maxRetries, fn, retryableException, retryCount -> computeSleepTimeMillis(retryCount, 50, 1000));
+        return future;
+    }
+
+    /**
+     * This is the same as {@link #retryWithBackoff(int, Supplier, Predicate)}, but takes a blocking retryable action
+     * and blocks the caller until done.
+     */
+    public static <A> A retryWithBackoffBlocking(final int maxRetries, final Supplier<A> fn)
+    {
+        return retryWithBackoffBlocking(maxRetries, fn, (ignore) -> true);
+    }
+
+    /**
+     * This is the same as {@link #retryWithBackoff(int, Supplier, Predicate)}, but takes a blocking retryable action
+     * and blocks the caller until done.
+     */
+    public static <A> A retryWithBackoffBlocking(final int maxRetries,
+                                                 final Supplier<A> fn,
+                                                 final Predicate<Throwable> retryableException)
+    {
+        return retryWithBackoff(maxRetries, () -> CompletableFuture.completedFuture(fn.get()), retryableException).join();
+    }
+
+    private static <A> void retryWithBackoff0(final CompletableFuture<A> result,
+                                              final int retryCount,
+                                              final int maxRetry,
+                                              final Supplier<CompletableFuture<A>> body,
```

Comment:
can we pass `Supplier<A> fn` directly? it might simplify a bit..

Reply:
this is the private implementation, so can't.  The `retryWithBackoffBlocking` calls use `Supplier<A>`  where as `retryWithBackoff` doesn't and uses a `Supplier<CompletableFuture<A>>`